nuxeo-compound-documents/src/main/java/org/nuxeo/compound/documents/CompoundDocumentServiceImpl.java

```diff
@@ -45,25 +52,48 @@
     @Override
     public DocumentModel createCompoundDocument(CoreSession session, String parent, Blob archive) {
         String compoundDocName = FilenameUtils.removeExtension(archive.getFilename());
-        DocumentModel compoundDoc = session.createDocumentModel(parent, compoundDocName, COMPOUND_DOCTYPE);
-        compoundDoc.setPropertyValue("dc:title", compoundDocName);
-        compoundDoc = session.createDocument(compoundDoc);
-        createCompoundDocument(compoundDoc, archive);
-        return compoundDoc;
+        var parentDoc = session.getDocument(new PathRef(parent));
+        try (var zip = new ZipFile(archive.getFile())) {
+            Set<String> entries = zip.stream().map(ZipEntry::getName).collect(Collectors.toSet());
+            String compoundType = runScript(parentDoc, COMPOUND_DOCTYPE_DETECTION_OPERATION, Map.of("entries", entries));
+            DocumentModel compoundDoc = session.createDocumentModel(parent, compoundDocName, compoundType);
+            compoundDoc.setPropertyValue("dc:title", compoundDocName);
+            compoundDoc = session.createDocument(compoundDoc);
+            createCompoundDocument(compoundDoc, zip);
+            return compoundDoc;
+        } catch (OperationException e) {
+            var message = String.format("Could not determine the CompoundDocument type for archive: %s.",
+                    archive.getFilename());
+            throw new NuxeoException(message, e);
+        } catch (IOException e) {
+            var message = String.format("Failed to create CompoundDocument: %s. Something went wrong with archive: %s.",
+                    compoundDocName, archive.getFilename());
+            throw new NuxeoException(message, e);
+        }
     }
 
-    protected void createCompoundDocument(DocumentModel compoundDoc, Blob archive) {
-        try (ZipFile zip = new ZipFile(archive.getFile())) {
-            zip.stream()
-               .filter(this::isAllowedEntry)
-               .sorted(Comparator.comparing(ZipEntry::getName))
-               .forEach(entry -> createEntry(compoundDoc, archive.getFilename(), zip, entry));
-        } catch (IOException e) {
-            String message = String.format(
-                    "Failed to create CompoundDocument: %s. something went wrong with archive: %s.",
-                    compoundDoc.getName(), archive.getFilename());
+    protected String runScript(DocumentModel doc, String scriptId, Map<String, ?> params)
+            throws OperationException {
+        AutomationService automationService = Framework.getService(AutomationService.class);
+        try (OperationContext ctx = new OperationContext()) {
+            ctx.setInput(doc);
+            return (String) automationService.run(ctx, scriptId, params);
```

Comment:
`null` would be infered as a string so it will be caught later:
```
java.lang.IllegalArgumentException: null is not a registered core type
	at org.nuxeo.ecm.core.api.DocumentModelFactory.createDocumentModel(DocumentModelFactory.java:167)
	at org.nuxeo.ecm.core.api.AbstractSession.createDocumentModelFromParentAndType(AbstractSession.java:679)
	at org.nuxeo.ecm.core.api.AbstractSession.createDocumentModelFromTypeName(AbstractSession.java:674)
	at org.nuxeo.ecm.core.api.AbstractSession.createDocumentModel(AbstractSession.java:707)
	at org.nuxeo.compound.documents.CompoundDocumentServiceImpl.createEntry(CompoundDocumentServiceImpl.java:112)
	at org.nuxeo.compound.documents.CompoundDocumentServiceImpl.lambda$3(CompoundDocumentServiceImpl.java:96)```

`Null` will trigger the expected mechanism, a caught NuxeoException with a specific message:
```org.nuxeo.ecm.core.api.NuxeoException: Could not determine the CompoundDocumentFolder type for compound document: test.
	at org.nuxeo.compound.documents.CompoundDocumentServiceImpl.createCompoundDocument(CompoundDocumentServiceImpl.java:91)
	at org.nuxeo.compound.documents.CompoundDocumentServiceImpl.createCompoundDocument(CompoundDocumentServiceImpl.java:62)
	at org.nuxeo.compound.documents.CompoundDocumentArchiveImporter.createOrUpdate(CompoundDocumentArchiveImporter.java:33)
	at org.nuxeo.ecm.platform.filemanager.service.FileManagerService.createOrUpdateDocument(FileManagerService.java:362)
	at org.nuxeo.compound.documents.TestCompoundDocumentImporter.testCustomCompoundDocumentType(TestCompoundDocumentImporter.java:90)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)```

Reply:
So worth catching the `null` early on in this method to throw a meaningful exception, right?

Reply:
The null check is not taken into account.

Reply:
Sorry didn't get the whole point. Now everything makes more sense !